lib/Dialect/Substrait/Transforms/EmitDeduplication.cpp

```diff
@@ -166,14 +220,61 @@ struct PushDuplicatesThroughCrossPattern : public OpRewritePattern<CrossOp> {
   }
 };
 
+/// Pushes duplicates in the mappings of `emit` ops producing the input through
+/// the `filter` op. This works by introducing a new `emit` op without the
+/// duplicates, creating a new `filter` op updated to work on the deduplicated
+/// element type, and finally a new `emit` op that maps back to the original
+/// order.
+struct PushDuplicatesThroughFilterPattern : public OpRewritePattern<FilterOp> {
+  using OpRewritePattern<FilterOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(FilterOp op,
+                                PatternRewriter &rewriter) const override {
+    if (!isa_and_present<EmitOp>(op.getInput().getDefiningOp()))
```

Comment:
Should we dyn_cast here and reuse result below?

Reply:
Yep -- much better! Done.